### PR TITLE
refactor: make the Entry struct more idiomatic Rust

### PR DESCRIPTION
**Pull request by Boshen**

### CODE REVIEW

File: src/entry.rs [hunk 1]
```diff
@@ -220,24 +220,12 @@ impl Resolver {
...
-            if let Some(kind) = kind {
-                if kind.is_file() || kind.is_dir() {
-                    file_dependency.push(entry.path.clone());
-                } else {
-                    miss_dependency.push(entry.path.clone());
-                }
-            }
-        }
-        (file_dependency, miss_dependency)
+        todo!("get_dependency_from_entry")
```
Comment (Boshen) [inline src/entry.rs:223]: This method is broken, and it's not used in rspack right now, so I removed it for less confusion.

File: src/entry.rs [hunk 2]
```diff
@@ -12,61 +11,69 @@ use crate::{
...
 
     pub fn is_dir(&self) -> bool {
-        matches!(self, EntryKind::Dir)
+        matches!(self.kind, EntryKind::Dir)
     }
 
-    pub fn exist(&self) -> bool {
-        matches!(self, EntryKind::NonExist)
+    pub fn exists(&self) -> bool {
+        !matches!(self.kind, EntryKind::NonExist)
```
Comment (bvanjoi) [inline src/entry.rs:49]: maybe sometime `self.kind` is `Unkown`.

File: src/entry.rs [hunk 3]
```diff
@@ -12,61 +12,50 @@ use crate::{
...
-    Dir,
-    NonExist,
-    Unknown,
+pub struct EntryStat {
+    /// `None` for non-existing file
+    file_type: Option<FileType>,
+
+    /// `None` for existing file but without system time.
+    modified: Option<SystemTime>,
 }
```
Comment (Boshen) [inline src/entry.rs:21]: I just moved FileType from `path.metadata()` here.

File: src/entry.rs [hunk 4]
```diff
@@ -12,61 +12,50 @@ use crate::{
...
 #[derive(Debug)]
 pub struct Entry {
-    pub parent: Option<Arc<Entry>>,
-    pub path: PathBuf,
-    pub pkg_info: Option<Arc<PkgInfo>>,
-    pub stat: RwLock<Option<EntryStat>>,
+    parent: Option<Arc<Entry>>,
+    path: Box<Path>,
+    pkg_info: Option<Arc<PkgInfo>>,
+    stat: RwLock<Option<Arc<EntryStat>>>,
```
Comment (bvanjoi) [inline src/entry.rs:58]: why wrapped by `Arc`?

File: src/entry.rs [hunk 5]
```diff
@@ -11,62 +11,51 @@ use crate::{
     RResult, Resolver,
 };
 
-#[derive(Debug, Clone)]
-pub enum EntryKind {
-    File,
-    Dir,
-    NonExist,
-    Unknown,
+#[derive(Debug, Clone, Copy)]
```
Comment (Boshen) [inline src/entry.rs:14]: We now implement `Copy`!

File: src/entry.rs (hunk 4 above)
Comment (Boshen) [reply src/entry.rs]: No longer needed, EntryStat is `Copy` now.

**Comment by bvanjoi** [comment]
thanks for your great work

**Review by bvanjoi — APPROVED** [review]

File: src/entry.rs [hunk 6]
```diff
@@ -11,62 +11,51 @@ use crate::{
...
+pub struct EntryStat {
+    /// `None` for non-existing file
+    file_type: Option<FileType>,
```
[the remark this answers is not in the document]
Comment (Boshen) [reply src/entry.rs:17]: Nope. See https://doc.rust-lang.org/std/fs/struct.FileType.html#method.is_symlink

> The underlying [Metadata](https://doc.rust-lang.org/std/fs/struct.Metadata.html) struct needs to be retrieved with the [fs::symlink_metadata](https://doc.rust-lang.org/std/fs/fn.symlink_metadata.html) function and not the [fs::metadata](https://doc.rust-lang.org/std/fs/fn.metadata.html) function. The [fs::metadata](https://doc.rust-lang.org/std/fs/fn.metadata.html) function follows symbolic links, so [is_symlink](https://doc.rust-lang.org/std/fs/struct.FileType.html#method.is_symlink) would always return false for the target file.

**Comment by Boshen** [comment]
I can finally move onto porting the DescriptionFilePlugin lol

**Pull request merged**